validator.go

```diff
@@ -109,3 +106,21 @@ func newValidatorImpl(domains []string, usersFile string,
 func NewValidator(domains []string, usersFile string) func(string) bool {
 	return newValidatorImpl(domains, usersFile, nil, func() {})
 }
+
+// isEmailValidWithDomains checks if the authenticated email is validated against the provided domain
+func isEmailValidWithDomains(email string, emailDomains []string) bool {
+	for _, domain := range emailDomains {
+		// allow if the domain is perfect suffix match with the email
+		if strings.HasSuffix(email, "@"+domain) {
+			return true
+		}
+
+		// allow if the domain is prefixed with . and
+		// the last element (split on @) has the suffix as the domain
+		atoms := strings.Split(email, "@")
+		if strings.HasPrefix(domain, ".") && strings.HasSuffix(atoms[len(atoms)-1], domain) {
```

Comment:
This implementation looks like it has a security flaw:

If you only allow `.mycompany.com`, if I register `.evilhackmycompany.com` it will pass that check due to the prefix `.` && suffix `mycompany.com` check.

We've had this vulnerability before (...maybe even here). Let me look up the CVE.

Reply:
https://github.com/oauth2-proxy/oauth2-proxy/security/advisories/GHSA-4mf2-f3wh-gvf2

Reply:
Oh wait -- maybe this is fine?  I mixed up the variable name `domain` a lot in my head while reading since it is a bit overloaded in this context 😅 

Can you potentially make this more clear -- maybe `allowedDomains` instead of `emailDomains`, so we can distinguish the value to check (user's email domain) vs the allowlist?

More logic/naming clarity will help avoid future contributors from introducing security bugs (probably the root of the open redirect CVE over time because that method is a nightmare).

Reply:
If it isn't there already -- can you add the attack technique I mentioned above to the unit tests to make sure it isn't valid?

Reply:
Hey @NickMeves , 
I've renamed the variable and also added two more tests depicting the attack you mentioned. 
All good

Reply:
Looking at the code, I agree, this is currently vulnerable. You need to split check for equality not a suffix in this case else you could have `foo@evilcompany.com` allowed right now when you intend to allow `foo@good.company.com`. I think you'll want to split on `.` too and check the last atom of this split for equality

Reply:
@JoelSpeed well, the first part of that `if` statement checks if the allowed domain starts with `.`  which means, in this scenario, we only allow subdomains/domains exact matches. Meaning the `evilcompany.com` wouldn't match a `company.com` because it evaluates to false on the first part of the if statement.

I've added an additional two test cases to check for this scenario and they show that the code is OK.
```
		{
			name:           "CheckForEqualityNotSuffix",
			email:          "foo@evilcompany.com",
			allowedEmails:  []string(nil),
			allowedDomains: []string{".company.com"},
			expectedAuthZ:  false,
		},
		{
			name:           "CheckForEqualityNotSuffix2",
			email:          "foo@evilcompany.com",
			allowedEmails:  []string(nil),
			allowedDomains: []string{"company.com"},
			expectedAuthZ:  false,
		},
```

Reply:
I think he fell for the same naming confusion I did 😄 

explicitly naming that `allowedDomain` instead of `domain` and getting a more specific name for the email domain parts than `atom` might help disambiguate the areas of concern.